.circleci/ci-helper.sh

```diff
@@ -1,13 +1,75 @@
+#!/usr/bin/env bash
 # inspired by https://github.com/Originate/guide/blob/master/android/guide/Continuous%20Integration.md
 
+function printTestsToRun {
+    if [ -z "$CIRCLE_PULL_REQUEST" ]; then
+
+        echo "Not a PR.  Run everything"
+    else
+        LIBS_TO_TEST=$(ruby .circleci/gitChangedLibs.rb)
+        echo -e "export LIBS_TO_TEST=${LIBS_TO_TEST}" >> "${BASH_ENV}"
+        if [[ ! -z ${LIBS_TO_TEST} ]]; then
+            echo -e "\n\nLibraries to Test-> ${LIBS_TO_TEST//","/", "}."
+        else
+            echo -e "\n\nNothing to Test."
+        fi
+    fi
+}
+
+function startAVD {
+    export LD_LIBRARY_PATH=${ANDROID_HOME}/emulator/lib64:${ANDROID_HOME}/emulator/lib64/qt/lib
+
+    # This indicates a nightly build and what API version to test
+    if [ -z "$AVD" ]; then
+        if [ -z "$CIRCLE_PULL_REQUEST" ] || [[ ${LIBS_TO_TEST} == *"${CURRENT_LIB}"* ]]; then
+            emulator64-arm -avd test22 -noaudio -no-window -accel on
+        else
+            echo "No need to start an emulator to test ${CURRENT_LIB} for this PR."
+        fi
+    else
+        emulator -avd "$AVD" -no-audio -no-window -accel on
+    fi
+}
+
 function waitForAVD {
-  local bootanim=""
-  export PATH=$(dirname $(dirname $(which android)))/platform-tools:$PATH
-  until [[ "$bootanim" =~ "stopped" ]]; do
-    sleep 5
-    bootanim=$(adb -e shell getprop init.svc.bootanim 2>&1)
-    echo "emulator status=$bootanim"
-  done
-  sleep 30
-  echo "Device Booted"
+    set +e
+
+    if [ -z "$CIRCLE_PULL_REQUEST" ] || [[ ${LIBS_TO_TEST} == *"${CURRENT_LIB}"* ]]; then
+        local bootanim=""
+        export PATH=$(dirname $(dirname $(which android)))/platform-tools:$PATH
+        until [[ "$bootanim" =~ "stopped" ]]; do
+            sleep 5
+            bootanim=$(adb -e shell getprop init.svc.bootanim 2>&1)
+            echo "emulator status=$bootanim"
+        done
+        sleep 30
+        # unlock the emulator screen
+        adb shell input keyevent 82
+        echo "Device Booted"
+    else
+        echo "No need to start an emulator to test ${CURRENT_LIB} for this PR."
+    fi
 }
+
+function runTests {
+    if [ -z "$CIRCLE_PULL_REQUEST" ] || [[ ${LIBS_TO_TEST} == *"${CURRENT_LIB}"* ]]; then
+        ./gradlew :libs:${CURRENT_LIB}:connectedAndroidTest --continue --no-daemon --profile --max-workers 2 --stacktrace
+    else
+        echo "No need to run ${CURRENT_LIB} tests for this PR."
+    fi
+}
+
+function runDanger {
+    if [ -z "$CIRCLE_PULL_REQUEST" ] || [[ ${LIBS_TO_TEST} == *"${CURRENT_LIB}"* ]]; then
+        if [ -z "${CURRENT_LIB}" ]; then
+            DANGER_GITHUB_API_TOKEN="c21349d8a97e1bf9cdd9""301fd949a83db862216b" danger --dangerfile=.circleci/Dangerfile_PR.rb --danger_id=PR-Check --verbose
```

Comment:
Should the token be read from an env variable instead of being hardcoded and publicly visible here?

Reply:
@bhariharan Unfortunately, this needs to be hardcoded (and It has been this way for a few months now).  See my previous comment: https://github.com/forcedotcom/SalesforceMobileSDK-Android/pull/1545#discussion_r150395729